### ISSUE TITLE
Consteval method in C++20/Clang unhappy?

### ISSUE THREAD

**Issue by sudara** [sudara/melatonin_perfetto#13]
@benthevining I think we went a bit overboard when trying to solve the perfetto C++17 deprecation, maybe you could take a look?

I'm having trouble compiling my parent project in C++20, which has the following set:

```
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

It's only a problem in Release, not in Debug.

I thought maybe the problem was `CMAKE_CXX_STANDARD` in the main perfetto CMakeLists.txt and I should swap to 

```
# set a *minimum* of C++17, but allow higher
target_compile_features(perfetto PRIVATE cxx_std_17)
```

But that hasn't resolved the issue.

I get the following error (Release only):

```
clang: error: unable to execute command: Segmentation fault: 11
clang: error: clang frontend command failed due to signal (use -v to see invocation)
Apple clang version 14.0.0 (clang-1400.0.29.202)
Target: x86_64-apple-darwin22.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
clang: note: diagnostic msg: Error generating preprocessed source(s) - cannot generate preprocessed source with multiple -arch options.
[82/144] Building CXX object CMakeFiles/SineMachine_v11.dir/JUCE/modules/juce_core/juce_core.mm.o
```

This was also reproduced by someone on the forum.

**Comment by sudara** [comment]
Hm, I actually see the issue go away if I replace `MELATONIN_CONSTEVAL` with `constexpr`

Maybe having `consteval` in the file somehow insists we need C++20, but the project is insisting we compile at C++17? Not sure how to best inspect which standard is *actually* being used at compile time.

**Comment by sudara** [comment]
Just for the record, I've been blowing away my build directory to ensure there's no link issue due to cached build files...

**Comment by sudara** [comment]
I tried the following to get clang to tell me what it's being invoked with, but it didn't work:

```
set(CMAKE_VERBOSE_MAKEFILE ON CACHE BOOL "ON" FORCE)
```

Let me know if you have any ideas!

**Comment by sudara** [comment]
Ok, it worked to pass this to the generator command:

`-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON`

I could verify that 2 arches are being passed `-arch arm64 -arch x86_64`

This is because in release I have the following in my main CMakeLists.txt

```
if (CMAKE_BUILD_TYPE STREQUAL "Release")
    set(CMAKE_OSX_ARCHITECTURES arm64 x86_64)
endif ()
```

However, i still get a clang segfault when removing that code — it seems something else is wrong and the arch stuff is a red herring? 😭

**Comment by sudara** [comment]
In the end, could be a consteval/clang issue, there seem to be many frontend crashes related to consteval listed here: https://github.com/orgs/llvm/projects/14/views/1?filterQuery=consteval

**Comment by benthevining** [comment]
I think your change to setting the cxx_17 standard target property on the just the `perfetto` target should be fine. The `melatonin_perfetto` JUCE module also declares the minimum C++ standard as 17 in its module metadata block, so both the Projucer and the CMake function `juce_add_module` will handle that appropriately.

The multiple architectures thing seems to be a separate issue -- I have also noticed that Clang sometimes fails building universal binaries on my machine with other projects, but I usually get some undefined symbol linker errors, not an actual crash -- that's scary stuff!

One possible solution is to ditch `consteval` and simply declare that function `constexpr`.

Besides that, my best guess is that Clang is crashing because we're somehow invoking undefined behavior or doing something otherwise illegal in the consteval context, which Clang is forced to execute before parsing for correctness.

One thing that looks a bit suspicious is the `while (src[i] != '(')` -- I think we should be checking the indices here to make sure we don't overrun the string. I can submit a PR for this.

**Comment by sudara** [comment]
> I think we should be checking the indices here to make sure we don't overrun the string. I can submit a PR for this.

Damn, I had this thought too but still had the problem — but I wasn't checking `j` as well — that turned out to fix things, despite not making conceptual sense (`j` is always equal to or smaller than `i`!)

Lifesaver, thank you!

Yeah, this consteval stuff seems......... not really ready for prime time. No hints in the IDE, terrible error messages. I'll try to avoid in the future until it's more stable :D

**Comment by benthevining** [comment]
Yeah. I wish it had better tooling support.

So index-checking fixed the issue?

**Comment by sudara** [comment]
Ug, looks like not.... For some reason I thought I had a good compile, but still getting the clang crash... Interesting that you couldn't reproduce with RELEASE and `consteval`....

**Comment by benthevining** [comment]
Ah, I'm not sure I tried release, let me try that...

**Comment by benthevining** [comment]
How interesting...

**Comment by benthevining** [comment]
Oh sorry. I just merged the PR that fixed this.

**Comment by sudara** [comment]
Hmm, there's no fix for this one yet (checking for indexes won't work). The quick fix is just settle for `constexpr` I think...

**Comment by benthevining** [comment]
Oh whoops, I thought we had fixed this, sorry. 

I suppose we can always define `MELATONIN_CONSTEVAL` to `constexpr` if the compiler is Clang, or we can simply replace `MELATONIN_CONSTEVAL` with a literal `constexpr`.

**Comment by sudara** [comment]
Yeah, maybe literal constexpr is best for now... just has the downside where I'm not sure if it's running compile time or not.... would be nice to know if deriving those names is adding runtime overhead.

**Comment by benthevining** [comment]
Hmm, yeah... We could do something like this:
```cpp
template <auto N>
constexpr auto class_and_method_name (char const (&src)[N])
{
    if (! std::is_constant_evaluated())
        std::cout << "Parsing method name at runtime!" << std::endl;

    std::array<char, N> res = {};
    for (size_t i = 0; i < N; ++i)
    {
        // wait until after the return type (first space in the character)
        if (src[i] == ' ')
        {
            ++i; // skip the space
            size_t j = 0; // keep track of our source string

            // build result, stop when we hit the arguments
            while ((src[i] != '('))
            {
                res[j] = src[i];
                ++i; // increment character in source
                ++j; // increment character in result
            }
            return res;
        }
    }
    return res;
}
```
though of course, adding the extra IO to print that will introduce further overhead.

**Comment by benthevining** [comment]
Hmm... I guess Clang just doesn't want to evaluate that function at compile-time. I'm not sure how to work around this, it's possible Clang may have some sort of `permissive` flag that would convince it to do this at compile-time, but I'm not sure.

**Comment by sudara** [comment]
Maybe the problem is the fact that `perfetto::StaticString` isn't consteval friendly, or shouldn't be used.

When I tried to remove that I got a perfetto assert:

`/Users/sudara/projects/sinemachine/cmake-build-debug/_deps/perfetto-src/sdk/perfetto.h:17421:3: error: static_assert failed due to requirement 'IsValidEventNameType<char *, void>::value' "Event names must be static strings. To use dynamic event names, see https://perfetto.dev/docs/instrumentation/track-events#dynamic-event-names"`

The normal supported perfetto route is to pass a string literal like `TRACE_EVENT ("component", "myName")`

**Comment by sudara** [comment]
That seemed to pass the std::is_constant_evaluated() test but fail in perfetto...

**Comment by sudara** [comment]
options:

1. move to constexpr, don't worry about the disconcerting overhead of constantly deriving names at runtime (boo)
2. Remove TRACE_EVENT, TRACE_DSP and have people use TRACE_EVENT with literals (double boo)
3. Try to build a compile-time map of PRETTY_FUNCTION <-> clean names (might end up with same problem with perfetto digesting the strings)
4. Look into std::source_location for C++20
5. Look into having the macro output string literals to place in the perfetto macro (?)

**Comment by sudara** [comment]
Oh wait, looks like consteval works when perfetto::DynamicString is used. Then nothing is output at runtime, everything seems happy!

constexpr still happens at runtime....

With luck this'll fix #11 too...

**Comment by sudara** [comment]
Ah, but I get the good ole clang segfault 11 in Release.... back to the drawing board!

**Comment by benthevining** [comment]
Maybe we can use the `perfetto::DynamicString` solution, and mark it `consteval` unless the compiler is Clang

**Comment by sudara** [comment]
I might be too stubborn to not have a clang solution :D :D :D :D

**Comment by benthevining** [comment]
Yeah... unfortunately I'm not sure what the solution would be. It seems like Clang's consteval support is not entirely standards conformant.

**Comment by sudara** [comment]
Ok, this is sort of insane, but works: https://godbolt.org/z/rzrTWK51j

The premise is that you wrap a string literal in a lambda so that it can behave as a compile-time template parameter. That way there's one compile-time template created per call site (I think? I hope?).

The strategies are talked about in depth here: https://accu.org/journals/overload/30/172/wu/#_idTextAnchor004

So seems happy at compile time (as tested with `!std::is_constant_evaluated()`)

Some of the problems I uncovered:

* A constexpr function will happily evaluate at runtime unless something *else* that's constexpr is asking for it
* Perfetto *really* wants to see a string literal passed to `StaticString`
* The existing template was based on `N`, the character count of the input string, so I *think* that's why I would see de-unique'ing of strings — any two strings of the same length were using the same template.

Issues with the current approach:
* yikes?
* I've set `N` to a constant of 100 characters. This could go back to `N`, i guess, if we add it as a template parameter, essentially passing a wrapped and unwrapped version of the string.
* wow.
